### PR TITLE
Allow integrating data into IConfigManager

### LINKED ISSUE #710: Allow optional integration of configuration providers in MORYX ConfigurationManager
**Is your feature request related to a problem? Please describe.**
As described in #672 it is an anti-pattern to put secrets into unencrypted config files within the applications working directory. 
> Never store passwords or other sensitive data in source code or configuration files. Production secrets shouldn't be used for development or test. Secrets shouldn't be deployed with the app. Production secrets should be accessed through a controlled means:
_[Source](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-9.0&tabs=windows)_

Currently, the means of getting configurations **and** secrets into MORYX components is through the `IConfigManager`. This component does not allow any special handling of secrets though.

**Describe the solution you'd like**
The .net way of managing secrets is by using [configuration providers](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration-providers). e.g. the .net [secret manager](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-9.0&tabs=windows#secret-manager) tool for development purposes or the [Azure Key Vault configuration provider](https://learn.microsoft.com/en-us/aspnet/core/security/key-vault-configuration?view=aspnetcore-9.0) for production purposes (among others).  
We should offer means to reference settings from configuration providers via the `IConfigManager` component. 

Acceptance Criteria:
- Properties in config classes are populated with values from config providers during runtime
  - Properties marked with a `PasswordAttribute` should search for these settings by default
  - Further options to indicate a property to be populated from a config provider _can_ be evaluated
- Any in the program.cs added config provider can be used as a data source
- _To be discussed:_ Referencing properties from a config provider is done just like from the code, e.g. `["Movies:ServiceApiKey"]` [(more information)](https://learn.microsoft.com/en-us/aspnet/core/security/app-secrets?view=aspnetcore-9.0&tabs=windows#read-the-secret-via-the-configuration-api)

**Describe alternatives you've considered**
#672 describes an alternative for securing app secrets based on an encryption mechanism for the whole MORYX configuration with a key stored as an environment variable.  
We propose this alternative solution for the following reasons:
- It is the way .net proposes to handle secrets
- It allows for very simple replacements of the secret providers between dev and prod environments
- It allows for a wide range of secret managers to be included and integrated with MORYX
- It follows the .net Best Practice **Don't treat nonsecrets like secrets.** explained in the info box [here](https://learn.microsoft.com/en-us/azure/well-architected/security/application-secrets)

**Additional context**
- There is a configuration provider for environment variables described [here](https://learn.microsoft.com/en-us/dotnet/core/extensions/configuration-providers#environment-variable-configuration-provider)
- There are isntructions to use environment variables when hosting in IIS described [here](https://stackoverflow.com/questions/75174574/net-core-pass-configuration-section-from-iis)
- There are instructions on how all this would need to be migrated from .net framework described [here](https://stackoverflow.com/questions/75174574/net-core-pass-configuration-section-from-iis)

### PR DESCRIPTION
**Pull request by andreniggemann**
### Summary

Allow populating ModuleConfigs from IConfiguration entries.
This is our way to allow injecting secrets and configuration from ConfigurationProviders based on Microsoft.Extensions.Configuration into the MORYX IConfigManager.

To properly handle nested configs (an important part of the IConfiguration concept the ValueProviderExecutor has been extended to allow access to context information through the new IContextAwareValueProvider interface.

### Linked Issues

CLoses #710 
It does not depend on the Password Attribute as mentionend in the Issue, because the Password Attribute still means that the value will be serialized to disk and over the API.

### Checklist for Submitter

- [x] I have tested these changes locally
- [x] I have updated documentation as needed
- [x] I have added or updated tests as appropriate
- [x] I have used clear and descriptive commit messages

### Review

**Typical tasks**

- [x] Merge request is well described
- [ ] Added `Obsolete` attributes if necessary
- [x] Critical sections are *documented in code*
- [x] *Tests* available or extended
- [x] *Documentation* is created / updated
- [x] Running in test environment
- [ ] *Manual* is created / updated
- [ ] Ports to other maintained versions are created

**Clean code**

- [x] *All* unused references are removed
- [x] Clean code rules are respected with passion (naming, ...)
- [x] Avoid *copy and pasted* code snippets

### CODE REVIEW

File: src/Moryx/Configuration/ValueProvider/IContextAwareValueProvider.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+// Licensed under the Apache License, Version 2.0
+
+using System.Reflection;
+
+namespace Moryx.Configuration;
+
+/// <summary>
+/// Stackframe to keep track of a recursive calls to the value provider to provide context.
+/// </summary>
+/// <param name="parent"></param>
```
Comment (1nf0rmagician) [inline src/Moryx/Configuration/ValueProvider/IContextAwareValueProvider.cs]: REmonde to fully populate

File: src/Moryx/Configuration/InjectedConfigAttribute.cs [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+/// <code>
+/// [InjectedConfig(".:ApiKey")]
+/// </code>
+/// Load the actual config key from the `ConfigKey` property
+/// <code>
+/// [InjectedConfig(nameof(ConfigKey), fromProperty: true)]
+/// </code>
+/// </example>
+[AttributeUsage(AttributeTargets.Property | AttributeTargets.Class)]
+public class InjectedConfigAttribute(string configKey, bool fromProperty = false) : Attribute
```
Comment (1nf0rmagician) [inline src/Moryx/Configuration/InjectedConfigAttribute.cs]: Suggestion to change to `ProvidedConfigAttribute` similar to the IConfiguration**Provider** which is used to provide the config?

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerTests.cs [hunk 3]
```diff
@@ -139,9 +151,170 @@ public void ByName()
...
+            name: "SecretConfig",
+            getCopy: true);
+
+        // Assert
+
+        Assert.That(config.NormalSetting, Is.EqualTo("from-json"));
+        Assert.That(config.ServiceApiKey, Is.EqualTo("from-provider"));
+    }
+
+    [Test(Description = "Resolves an explicit configuration key placeholder from IConfiguration.")]
```
Comment (1nf0rmagician) [inline src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerTests.cs]: copy paste

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerTests.cs [hunk 4]
```diff
@@ -139,9 +151,170 @@ public void ByName()
...
+        var writtenConfig = File.ReadAllText(configPath);
+
+        using var _ = Assert.EnterMultipleScope();
+        Assert.That(config.ServiceApiKey, Is.EqualTo("from-provider"));
+        Assert.That(writtenConfig, Does.Not.Contain("from-provider"));
+    }
+
+
+    [Test(Description = "Applies provider values to generated configs when no JSON file exists.")]
+    public void GeneratedConfig_ReceivesValuesFromConfiguration()
```
Comment (1nf0rmagician) [inline src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerTests.cs]: MIxed naming conventions, either keep it with the existing tests or apply the .net guideline consistently for new tests

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/MsExtensionConfigProviderTests.cs [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+        // Act
+        var result = provider.Handle(stack);
+
+        // Assert
+        using var _ = Assert.EnterMultipleScope();
+        Assert.That(result, Is.EqualTo(ValueProviderResult.Handled));
+        Assert.That(testConfig.ServiceApiKey, Is.EqualTo(Expected));
+    }
+
+    private static IConfigurationRoot BuildConfig(Dictionary<string, string> configDict)
```
Comment (1nf0rmagician) [inline src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/MsExtensionConfigProviderTests.cs]: could be used in other test class as well

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/MsExtensionConfigProviderTests.cs [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+        // Assert
+        using var _ = Assert.EnterMultipleScope();
+        Assert.That(result, Is.EqualTo(ValueProviderResult.Handled));
+        Assert.That(testConfig.ServiceApiKey, Is.EqualTo(Expected));
+    }
+
+    [TestCase(0)]
+    [TestCase(1)]
+    [TestCase(2)]
+    [TestCase(3)]
```
Comment (1nf0rmagician) [inline src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/MsExtensionConfigProviderTests.cs]: I think 2 test cases are enough here 🤣

File: src/Moryx.Runtime.Kernel/Configuration/ConfigManager.cs [hunk 7]
```diff
@@ -34,8 +38,9 @@ public ConfigManager()
     protected IValueProvider[] ValueProviders =>
     [
         _sharedProvider,
+        new ActivatorValueProvider(),
```
Comment (1nf0rmagician) [inline src/Moryx.Runtime.Kernel/Configuration/ConfigManager.cs]: I assume this change in the order of the collection is what you are refering to in the description, which would need a proper testing for its side effects?

File: src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+    public ValueProviderResult Handle(Stack<ExecutorLevel> levels)
+    {
+        var (config, prop, dict) = levels.Peek();
+        if (_configuration == null || config == null || !prop.CanRead || !prop.CanWrite)
+        {
+            return ValueProviderResult.Skipped;
+        }
+        var currentValue = prop.GetValue(config);
+        var defaultValue = GetDefaultValue(prop.PropertyType);
+        if (!Equals(currentValue, defaultValue)) // value is already assigned
```
Comment (1nf0rmagician) [inline src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs]: For Meaning, the intention behind the usage of configuration providers with MORYX 10 is to initialize or to provide not externally visible config values. We do not store values back to any configuration provider and we skip any application changed properties.

Should we have a test case to document this as intended behaviour?

File: src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs [hunk 9]
```diff
@@ -0,0 +1,143 @@
...
+        {
+            return ValueProviderResult.Skipped;
+        }
+
+
+        var configKey = GetConfigKeyNameCached(levels);
+
+        if (_configuration.GetSection(configKey) is ConfigurationSection section && section.Exists())
+        {
+            prop.SetValue(config, section.Get(prop.PropertyType));
```
Comment (1nf0rmagician) [inline src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs]: This would throw if the config contains a wrong datatype, correct? Should be catched with an understandable error message then (and tested)

File: src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs [hunk 10]
```diff
@@ -0,0 +1,143 @@
...
+            var baseKey = GetRootConfigKey(levels);
+            return $"{baseKey}:{key}";
+        }
+        
+    }
+
+    private static string KeyFromAttribute(object config, InjectedConfigAttribute attribute)
+    {
+        var configKey = attribute.ConfigKey;
+        if (attribute.FromProperty)
```
Comment (1nf0rmagician) [inline src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs]: You could throw nice exceptions if someone missconfigured the property name or the property is not a string property. I think currently the behaviour is rather undefined

File: src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+            if (!(property is null || property.PropertyType != typeof(string) || !property.CanRead))
+            {
+                configKey = property.GetValue(config) as string;
+            }
+        }
+
+        return configKey;
+    }
+
+    private object GetRootConfigKey(Stack<ExecutorLevel> levels)
```
Comment (1nf0rmagician) [inline src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs]: This should return string shouldn'T it?

File: src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+        if (_configuration.GetSection(configKey) is ConfigurationSection section && section.Exists())
+        {
+            prop.SetValue(config, section.Get(prop.PropertyType));
+            return ValueProviderResult.Handled;
+        }
+
+        return ValueProviderResult.Skipped;
+    }
+
+    private string GetConfigKeyNameCached(Stack<ExecutorLevel> levels)
```
Comment (1nf0rmagician) [inline src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs]: I must admit I do not quite get the caching part in this, maybe a little explanation comment/example comment would help me🙈

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/MsExtensionConfigProviderTests.cs
Comment (1nf0rmagician) [inline src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/MsExtensionConfigProviderTests.cs:1]: While you have implemented collection handling (as far as i can see) there is no test for this

**Review by 1nf0rmagician — COMMENTED** [review]
I'm happy with the concept and think it is a nicely limited change!

I already went quite into the details with the review, but please take this not as a being overly picky with the PR. I needed to read it closely for full understanding and when I spotted something I didn't want to forget about it later when you had brought it out of the draft state 😬🤗

**Review by dbeuchler — COMMENTED** [review]
I like that implementation! Thanks for that. (just small things from my end: code format, braces, no one liners...).

One thing to think about: Why do you have touched the default `ConfigManager`? I think this can easily archived by derive from it and provide an additional implementation which the application developer can select if required. Keeps the default implementation as it is and for extended usage, it is possible to use it in application.

Before we merge it to dev and make an release instantly, could we apply this into an application (I don't mean the demo) to test if it behaves as wanted?

**Comment by andreniggemann** [comment]
> I like that implementation! Thanks for that. (just small things from my end: code format, braces, no one liners...).

Thank you very much :)

> One thing to think about: Why do you have touched the default `ConfigManager`? I think this can easily archived by derive from it and provide an additional implementation which the application developer can select if required. Keeps the default implementation as it is and for extended usage, it is possible to use it in application.

That's good idea. I would add a switch to the AddMoryxKernel Extension method that allows opting in to the new behavior, if that's ok with you

> Before we merge it to dev and make an release instantly, could we apply this into an application (I don't mean the demo) to test if it behaves as wanted?

That's my plan. That's why it's still in draft. Additionally I intend to add a regression test that takes the ModuleConfigs of the core Modules and checks that, if the IConfiguration is empty, you get the exact same configs.

**Comment by dbeuchler** [comment]
> That's good idea. I would add a switch to the AddMoryxKernel Extension method that allows opting in to the new behavior, if that's ok with you

Instead of adding options to the extensions, another idea: You can replace registered services with `services.Replace`? Maybe a simple hint in the docs and this behavior is fully optional.

**Comment by andreniggemann** [comment]
Isn't it completely optional if I define the Method as `AddMoryxKernel(bool configurationIntegration = false)`?

Nobody would have to change their code, it's much more discoverable as a hint in the docs and the "new ConfigManager" (not sure on naming yet, or if even needs a new type. Maybe it could just be a different constructor that takes a list of IValueProviders) can be internal. 

By the way: Is there a reason for the existing ConfigManager being public? In an application we always use it through the IConfigManager interface

File: src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs (hunk 12 above)
Comment (andreniggemann) [reply src/Moryx.Runtime.Kernel/Configuration/ValueProvider/MsExtensionsConfigProvider.cs]: It only stores the config key for an element it has already seen in the dict attached to the stackframe.
This means that if we have nested configuration, you don't have to build the entire path from scratch each time.

File: docs/articles/framework/configuration.md [hunk 13]
```diff
@@ -64,11 +64,137 @@ In the configuration editor each entry is represented by its property name and v
...
+### Example
+The ModuleConfig of the ProductManagement module contains a property called MaxImporterWait.
+To configure this value via IConfiguration, define the following key:
+```
+Moryx.Products.Management.ModuleConfig:MaxImporterWait
+```
+This can be set in appsettings.json like this:
+
+
+You can archive this by adding the following to your appsettings.json
```
Comment (1nf0rmagician) [inline docs/articles/framework/configuration.md]: little duplicate

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerWithConfigProviderTests.cs [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+
+        var provider = serviceCollection.BuildServiceProvider();
+
+        var manager = provider.GetRequiredService<ConfigManager>();
+        manager.ConfigDirectory = _tempDir;
+
+        return manager;
+    }
+
+    // TODO: Consider removing -> Tested by MicrsoftExtensionConfigProviderTests
```
Comment (1nf0rmagician) [inline src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerWithConfigProviderTests.cs]: Open todos in this file, if you ask me, I like keeping some integration tests (if they actually test just a littel bit more than the `MicrsoftExtensionConfigProviderTests`.

**Review by 1nf0rmagician — DISMISSED** [review]

File: src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerWithConfigProviderTests.cs (hunk 14 above)
Comment (andreniggemann) [reply src/Tests/Moryx.Runtime.Kernel.Tests/Configuration/ConfigManagerWithConfigProviderTests.cs]: Ok, than I'll remove the TODOs. They test the integrate a little bit

File: src/Moryx.Runtime.Kernel/KernelServiceCollectionExtensions.cs [hunk 15]
```diff
@@ -20,10 +21,27 @@ public static class KernelServiceCollectionExtensions
...
+                {
+                    var manager = ActivatorUtilities.CreateInstance<ConfigManager>(serviceProvider);
+                    return manager.ConfigureValueProviders(original => [
+                        manager._sharedProvider,
+                        new ActivatorValueProvider(),
+                        ActivatorUtilities.CreateInstance<MicrosoftExtensionsConfigProvider>(serviceProvider),
+                        new DefaultValueAttributeProvider(),
+                    ]);
+                });
+            }
```
Comment (dbeuchler) [inline src/Moryx.Runtime.Kernel/KernelServiceCollectionExtensions.cs:40]: Instead of define that again here, couldnt it be also an option of the config manager which you set here? The extension knows to much about the internals of the config manager just for the flag.

Comment (andreniggemann) [reply src/Moryx.Runtime.Kernel/KernelServiceCollectionExtensions.cs:40]: Note for everyone else: @dbeuchler and I discussed this on the phone.

Because of the cyclic dependency between the SharedConfigProvider and the ConfigManager and their difference in visibility, it's not easily possible to add a subclass of the ConfigManager that does all this configuration, so we are going to keep it as is.

On the Future Branch this can be resolved by just removing the SharedConfigProvider as it is basically unused.

**Review by dbeuchler — APPROVED** [review]
